relay/pkg/header/integration_test.go

```diff
@@ -0,0 +1,292 @@
+			expectedAddHeadersWithRetargetEvent,
+			actualAddHeadersWithRetargetEvent,
+		)
+	}
+
+	markNewHeaviestEvents := localChain.MarkNewHeaviestEvents()
+
+	expectedEventsCount = 1
+	actualEventsCount = len(markNewHeaviestEvents)
+	if expectedEventsCount != actualEventsCount {
+		t.Fatalf(
+			"unexpected number of mark new heaviest events:\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]\n",
+			expectedEventsCount,
+			actualEventsCount,
+		)
+	}
+
+	expectedMarkNewHeaviestEvent := &chainlocal.MarkNewHeaviestEvent{
+		AncestorDigest:    to32Bytes(7),
+		CurrentBestHeader: toBytes(7),
+		NewBestHeader:     toBytes(12),
+		Limit:             big.NewInt(6),
+	}
+	actualMarkNewHeaviestEvent := markNewHeaviestEvents[0]
+	if !reflect.DeepEqual(expectedMarkNewHeaviestEvent, actualMarkNewHeaviestEvent) {
+		t.Errorf(
+			"unexpected mark new heaviest event:\n"+
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedMarkNewHeaviestEvent,
+			actualMarkNewHeaviestEvent,
+		)
+	}
+
+	localChain.SetBestKnownDigest(to32Bytes(12))
+	time.Sleep(testRelayPushingSleepTime)
+
+	//************ Verify processing of the second batch ************
+	addHeadersEvents := localChain.AddHeadersEvents()
+
+	expectedEventsCount = 1
+	actualEventsCount = len(addHeadersEvents)
+	if expectedEventsCount != actualEventsCount {
+		t.Fatalf(
+			"unexpected number of add headers events:\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]\n",
+			expectedEventsCount,
+			actualEventsCount,
+		)
+	}
+
+	expectedAddHeadersEvent := &chainlocal.AddHeadersEvent{
+		AnchorHeader: toBytes(12),
+		Headers:      toBytes(13, 14, 15),
+	}
+	actualAddHeadersEvent := addHeadersEvents[0]
+	if !reflect.DeepEqual(expectedAddHeadersEvent, actualAddHeadersEvent) {
+		t.Errorf(
+			"unexpected add headers event:\n"+
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedAddHeadersEvent,
+			actualAddHeadersEvent,
+		)
+	}
+
+	addHeadersWithRetargetEvents = localChain.AddHeadersWithRetargetEvents()
+
+	expectedEventsCount = 2
+	actualEventsCount = len(addHeadersWithRetargetEvents)
+	if expectedEventsCount != actualEventsCount {
+		t.Fatalf(
+			"unexpected number of add headers with retarget events:\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]\n",
+			expectedEventsCount,
+			actualEventsCount,
+		)
+	}
+
+	expectedAddHeadersWithRetargetEvent = &chainlocal.AddHeadersWithRetargetEvent{
+		OldPeriodStartHeader: toBytes(8),
+		OldPeriodEndHeader:   toBytes(15),
+		Headers:              toBytes(16, 17),
+	}
+
+	actualAddHeadersWithRetargetEvent = addHeadersWithRetargetEvents[1]
+
+	if !reflect.DeepEqual(
+		expectedAddHeadersWithRetargetEvent,
+		actualAddHeadersWithRetargetEvent,
+	) {
+		t.Errorf(
+			"unexpected add headers with retarget event:\n"+
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedAddHeadersWithRetargetEvent,
+			actualAddHeadersWithRetargetEvent,
+		)
+	}
+
+	markNewHeaviestEvents = localChain.MarkNewHeaviestEvents()
+
+	expectedEventsCount = 2
```

Comment:
Worth putting a comment on why we have two events here. May be not obvious for someone not familiar with relay logic.